src/thermo/PengRobinson.cpp

```
@@ -0,0 +1,834 @@
+{
+    getEntropy_R_ref(sbar);
+    scale(sbar, sbar+m_kk, sbar, GasConstant);
+    double T = temperature();
+    double mv = molarVolume();
+    double vmb = mv - m_b_current;
+    double vpb2 = mv + (1 + M_SQRT2)*m_b_current;
+    double vmb2 = mv + (1 - M_SQRT2)*m_b_current;
+    double refP = refPressure();
+    double daAlphadT = daAlpha_dT();
+    double coeff1 = 0;
+    double den1 = 2 * M_SQRT2 * m_b_current * m_b_current;
+    double den2 = mv * mv + 2 * mv * m_b_current - m_b_current * m_b_current;
+
+    // Calculate sum(n_j (a alpha)_i, k * (1/alpha_k d/dT(alpha_k))) -> m_pp
+    // Calculate sum(n_j (a alpha)_i, k * (1/alpha_i d/dT(alpha_i))) -> m_tmpV
+    for (size_t k = 0; k < m_kk; k++) {
+        m_pp[k] = 0.0;
+        m_tmpV[k] = 0;
+        for (size_t i = 0; i < m_kk; i++) {
+            m_pp[k] += moleFractions_[i] * m_aAlpha_vec_Curr(k, i);
+            m_tmpV[k] += moleFractions_[i] * m_aAlpha_vec_Curr(k, i) *(m_dalphadT_vec_Curr[i] / m_alpha_vec_Curr[i]);
+        }
+        m_pp[k] = m_pp[k] * m_dalphadT_vec_Curr[k] / m_alpha_vec_Curr[k];
+    }
+
+
+    for (size_t k = 0; k < m_kk; k++) {
+        coeff1 = m_b_current * (m_pp[k] + m_tmpV[k]) - daAlphadT * m_b_vec_Curr[k];
+        sbar[k] += GasConstant * log(GasConstant * T / (refP * mv))
+                   + GasConstant
+                   + GasConstant * log(mv / vmb)
+                   + GasConstant * m_b_vec_Curr[k] / vmb
+                   - coeff1* log(vpb2 / vmb2) / den1
+                   - m_b_vec_Curr[k] * mv * daAlphadT / den2 / m_b_current;
+    }
+    calculatePressureDerivatives();
+    getPartialMolarVolumes(m_partialMolarVolumes.data());
+    for (size_t k = 0; k < m_kk; k++) {
+        sbar[k] -= m_partialMolarVolumes[k] * m_dpdT;
+    }
+}
+
+void PengRobinson::getPartialMolarIntEnergies(double* ubar) const
+{
+    // u_i = h_i - p*v_i
+    double* vbar = 0;
+    double p = pressure();
+    getPartialMolarEnthalpies(ubar);
+    getPartialMolarVolumes(vbar);
```

The lack of a test that covers this method is hiding the error in this code.

---

Fixed.